### PR TITLE
HIVE-26735: Ability to sort the data during rebalancing compaction

### PR DESCRIPTION
**Pull request by veghlaci05**
### What changes were proposed in this pull request?
The REBALANCE compaction now accepts an ORDER BY clause.

### Why are the changes needed?
 To be able to reorder the data during the compaction.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manually and through unit tests.

### CODE REVIEW

**Comment by TuroczyX** [comment]
Nice. Pretty big changeset :)

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java [hunk 1]
```diff
@@ -237,7 +248,11 @@ String build() {
       query.append(" temporary external");
     }
     if (operation == Operation.INSERT) {
-      query.append(" into");
+      if (compactionType.equals(CompactionType.REBALANCE)) {
+        query.append(" overwrite");
+      } else {
+        query.append(" into");
+      }
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:255]: Can we do this - 
`query.append(compactionType.equals(CompactionType.REBALANCE) ? "overwrite" : "append");`

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestWorker.java [hunk 2]
```diff
@@ -1180,19 +1180,19 @@ private void verifyTxn1IsAborted(int compactionNum, Table t, CompactionType type
   }
 
   // With high timeout, but fast run we should finish without a problem
-  @Test(timeout=1000)
+  @Test(timeout=2000)
```
Comment (SourabhBadhya) [inline ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestWorker.java:1183]: Just curious, why is this changed?

Comment (veghlaci05) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestWorker.java:1183]: In one of the builds this test failed with timeout. Could not reproduce it locally, but increased the value in order to prevent flakyness in the CI builds.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java (hunk 1 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:255]: `CompactionType.REBALANCE.equals(compactionType)` would be null safe too.

Or equality operator is also fine because `CompactionType` is an enum.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 3]
```diff
@@ -1561,7 +1561,7 @@ public void commitTxn(CommitTxnRequest rqst) throws NoSuchTxnException, TxnAbort
           assert true;
         }
 
-        if (txnType != TxnType.READ_ONLY && !isReplayedReplTxn) {
+        if (txnType != TxnType.READ_ONLY && !isReplayedReplTxn && txnType != TxnType.COMPACTION && txnType != TxnType.REBALANCE_COMPACTION) {
```
Comment (kasakrisz) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1564]: qq: Why compaction txn components should not be moved here? Does it handled by `CompactionTxnHandler` ?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java (hunk 1 above)
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:255]: Fixed.

**Review by kasakrisz — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/Compiler.java [hunk 4]
```diff
@@ -263,7 +263,7 @@ private void setLastReplIdForDump(HiveConf conf) throws HiveException, TExceptio
 
   private void openTransaction(TxnType txnType) throws LockException, CommandProcessorException {
     if (DriverUtils.checkConcurrency(driverContext) && startImplicitTxn(driverContext.getTxnManager()) &&
-        !driverContext.getTxnManager().isTxnOpen() && txnType != TxnType.COMPACTION) {
+        !driverContext.getTxnManager().isTxnOpen() && txnType != TxnType.COMPACTION && txnType != TxnType.REBALANCE_COMPACTION) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/Compiler.java:266]: could we extract the compaction part into a helper method `isCompactionTxn()`

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 5]
```diff
@@ -289,6 +289,7 @@ private boolean validateTxnList() throws CommandProcessorException {
           }
           // Since we're reusing the compiled plan, we need to update its start time for current run
           driverContext.getPlan().setQueryStartTime(driverContext.getQueryDisplay().getQueryStartTime());
+          driverContext.setRetrial(false);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:292]: why do we need this?

Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:292]: The retrial flag is needed to re-create the plan, but when the plan is re-created the flag needs to be cleared, otherwise the next query will fail. This is a bug I found by the testRebalanceCompactionWithParallelDelete() test.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnQueries.java [hunk 6]
```diff
@@ -61,7 +61,7 @@ public class TxnQueries {
                     "   \"CQ_WORKER_VERSION\" AS \"CC_WORKER_VERSION\", \"CQ_INITIATOR_ID\" AS \"CC_INITIATOR_ID\", \"CQ_INITIATOR_VERSION\" AS \"CC_INITIATOR_VERSION\", " +
                     "   \"CQ_RETRY_RETENTION\" AS \"CC_RETRY_RETENTION\", \"CQ_NEXT_TXN_ID\" AS \"CC_NEXT_TXN_ID\", \"CQ_TXN_ID\" AS \"CC_TXN_ID\", " +
                     "   \"CQ_COMMIT_TIME\" AS \"CC_COMMIT_TIME\", \"CQ_POOL_NAME\" AS \"CC_POOL_NAME\",  " +
-                    "   \"CQ_NUMBER_OF_BUCKETS\" AS \"CC_NUMBER_OF_BUCKETS\" " +
+                    "   \"CQ_NUMBER_OF_BUCKETS\" AS \"CC_NUMBER_OF_BUCKETS\", \"CQ_ORDER_BY_CLAUSE\" AS \"CC_ORDER_BY_CLAUSE\" " +
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnQueries.java:64]: maybe simply `CQ_ORDER_BY`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 7]
```diff
@@ -1561,7 +1561,7 @@ public void commitTxn(CommitTxnRequest rqst) throws NoSuchTxnException, TxnAbort
           assert true;
         }
 
-        if (txnType != TxnType.READ_ONLY && !isReplayedReplTxn) {
+        if (txnType != TxnType.READ_ONLY && !isReplayedReplTxn && txnType != TxnType.REBALANCE_COMPACTION) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1564]: why is this change?

Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1564]: why don't we do the same for TxnType.COMPACTION and instead have the if block `AND \"TC_OPERATION_TYPE\" <> " + OperationType.COMPACT` under moveTxnComponentsToCompleted?

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1564]: i think that is because we didn't have TxnType.COMPACTION type before the original change

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java [hunk 8]
```diff
@@ -87,6 +90,262 @@
 @SuppressWarnings("deprecation")
 public class TestCrudCompactorOnTez extends CompactorOnTezTest {
 
+
+  @Test
+  public void testRebalanceCompactionWithParallelDelete() throws Exception {
+    conf.setBoolVar(HiveConf.ConfVars.COMPACTOR_CRUD_QUERY_BASED, true);
+    conf.setBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_GATHER_STATS, false);
+    conf.setBoolVar(HiveConf.ConfVars.HIVESTATSAUTOGATHER, false);
+
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java:99]: please create a parametrized test and validate for `TXN_WRITE_X_LOCK=false` as well

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java (hunk 5 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:292]: 👍

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java (hunk 8 above)
Comment (deniskuzZ) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java:99]: that flag controls optimistic/pessimistic concurrency control mode

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 7 above)
Comment (veghlaci05) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1564]: Normal compaction is not blocking (Uses SHARED_READ lock) so there is no need for conflict detection. Rebalance compaction uses EXCL_WRITE lock, and the records must not moved to the complete table in order to be able to detect conflicting write attempts. This is required to force concurrent updates/deletes to re-create their plan and use the new base/deltas containing the changed ACID rows.

**Review by deniskuzZ — APPROVED** [review]
+1
